core/src/program/info.rs

```diff
@@ -0,0 +1,100 @@
+
+/// A program information set consisting of its code block root and set of kernel roots used for
+/// its compilation.
+///
+/// This will be used as public inputs of the proof so we bind its verification to the kernel and
+/// root used to execute the program. This way, we extend the correctness of the proof to the
+/// security guarantees provided by the kernel. We also allow the user to easily prove the
+/// membership of a given kernel procedure for a given proof, without compromising its
+/// zero-knowledge properties.
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+pub struct ProgramInfo {
+    program_hash: Digest,
+    kernel: Kernel,
+}
+
+impl ProgramInfo {
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Creates a new instance of a program info.
+    pub const fn new(program_hash: Digest, kernel: Kernel) -> Self {
+        Self {
+            program_hash,
+            kernel,
+        }
+    }
+
+    /// Creates a new program info instance from a program root and a kernel set.
+    pub fn from_program_parts<R>(program_hash: R, kernel: Kernel) -> Self
+    where
+        R: AsRef<CodeBlock>,
+    {
+        Self {
+            program_hash: program_hash.as_ref().hash(),
+            kernel,
+        }
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns the program hash computed from its code block root.
+    pub const fn program_hash(&self) -> &Digest {
+        &self.program_hash
+    }
+
+    /// Returns the program kernel used during the compilation.
+    pub const fn kernel(&self) -> &Kernel {
+        &self.kernel
+    }
+
+    /// Returns the list of procedures of the kernel used during the compilation.
+    pub fn kernel_procedures(&self) -> &[Digest] {
+        self.kernel.proc_hashes()
+    }
+}
+
+impl From<Program> for ProgramInfo {
+    fn from(program: Program) -> Self {
+        let Program { root, kernel, .. } = program;
+        let program_hash = root.hash();
+
+        Self {
+            program_hash,
+            kernel,
+        }
+    }
+}
+
+impl Serializable for ProgramInfo {
+    fn write_into(&self, target: &mut ByteWriter) -> Result<(), SerializationError> {
+        target.write_digest(&self.program_hash);
+        self.kernel.write_into(target)
+    }
+}
+
+impl Deserializable for ProgramInfo {
+    fn read_from(bytes: &mut ByteReader) -> Result<Self, SerializationError> {
+        let program_hash = bytes.read_digest()?;
+        let kernel = Kernel::read_from(bytes)?;
+        Ok(Self {
+            program_hash,
+            kernel,
+        })
+    }
+}
+
+impl winter_utils::Serializable for ProgramInfo {
+    fn write_into<W: winter_utils::ByteWriter>(&self, target: &mut W) {
+        self.program_hash.write_into(target);
+        <Kernel as winter_utils::Serializable>::write_into(&self.kernel, target);
+    }
+}
```

Comment:
Similar comment re serialization as before.

Reply:
Ack! Resolving this discussion to keep the topic centralized on the previous one